eip712/src/main/java/pm/gnosis/eip712/EIP712.kt

```diff
@@ -0,0 +1,51 @@
+package pm.gnosis.eip712
+
+import pm.gnosis.crypto.utils.Sha3Utils
+import pm.gnosis.model.Solidity
+import pm.gnosis.model.SolidityBase
+import pm.gnosis.utils.hexToByteArray
+
+val solidityTypes = Solidity.types.keys
+const val DOMAIN_TYPE = "EIP712Domain"
+
+sealed class Type712 {
+    abstract val type: String
+    abstract fun encodeData(): ByteArray
+}
+
+class Literal712(val value: SolidityBase.Type) : Type712() {
+    override fun encodeData(): ByteArray = when (value) {
+        // Solidity.String is also Solidity.Bytes
+        is Solidity.Bytes -> Sha3Utils.keccak(value.items)
+        is SolidityBase.Array<*> -> value.items.map { Literal712(value = it).encodeData() }.reduce { acc, bytes -> acc + bytes }
```

Comment:
These should be hashed again 
From the EIP:
```
The array values are encoded as the keccak256 hash of the concatenated encodeData of their contents (i.e. the encoding of SomeType[5] is identical to that of a struct containing five members of type SomeType).
```

Reply:
Good catch!